### PR TITLE
Demo Subwidget in hpp-gui

### PR DESCRIPTION
**Pull request by ymontmarin**
@jmirabel 
This pull request intend to add the work done in July to add tools to manipulate gepetto-gui + hpp-gui as a standalone demonstrator + minor fixes that I encounter when adding the features.

## Demo feature
The goal is to have tools to reset hpp and gepetto, load and save problem setup for given env/robot setup... To be robust, the feature need access to other widget for reseting, loading and saving. Globally it constitutes 3 buttons on the toolbar (reset / load /save) and to be the least intrusive possible, I add the feature in a subwidget in the separate file `demosubwidget.cc`. But add to slightly adapt the rest of the code:
- For the load/save function I needed to add TinyXML as a dependency in CMake.
- `reciveConfig`, `resetAllConfigs`, `reinitialize` methods in the `configurationlistwidget` allowing to load configuration in the list from a file and flush configurations of the list. A method `deleteAll` in `configurationlist.cc` to operate with the widget methods.
- in the main class `hppwidgetsplugin.cc`, attribute, init and delete of the demo subwidget

## Minor change
- fix index of default config name in `configurationlistwidget` to start with `config_0, config_1` instead of `config, config_0` which is easier when loading config and keep indexes.
- Use of `ConfigurationList` type instead of parent type `QListWidget` in method `list`
- in loading robot method of in `hppwidgetsplugin.cc` use the method `fetchConfiguration` instead of `requestApplyCurrentConfiguration` which is include in `fetchConfiguration`. It allow to load the robot directly in the neutral configuration without the need of a manual refresh.
- wrapping `prepareApplyConfiguration` in a qDebug message to avoid error when using refresh without loaded robot
- public method `configurationListWidget` in `hppwidgetsplugin.cc` to expose the configuration widget as done for other widget.
- use  qDebug instead of logError in `jointtreewidget.cc` to avoid error when using refresh without loaded robot
- in `roadmap.cc` (resp `pathplayer.cc`) delete and recompute roadmap (resp path) if exist to allow changing the roadmap (resp path) if a new solving has been done.
- in `roadmap.cc` add the roadmap to hpp-gui group only at initialization and not at each iteration which is not needed.

From my different testing, all seems to work fine, but I don't know if there is any test procedure. This merge could be useful for the use of the demonstrator without the need to compile a wild version of hpp-gui. I hope some of the changes can be useful for you. @jmirabel , let me know if there is some constrains i didn't take into account. Thank you !

PS: I am currently using the compilation I made of all laas package in conda env to build some official conda-forge feedstock for those. Hope it can help laas team and I'll keep you inform !

### CODE REVIEW

**Comment by hrp2-14** [comment]
This project doesn't accept pull requests on master, please change the base branch of your pull request to devel.

**Comment by jmirabel** [comment]
At the moment, and for about 2 weeks, I will not be able to work. Is this PR urgent ?
Otherwise, it will wait for me going back to work (or someone else from LAAS review it).

**Comment by ymontmarin** [comment]
@jmirabel No, the PR is not urgent yet, for the moment I use my custom build. The PR will be useful to efficiently install the demonstrator on other system e.g. for robotic courses but there is no special need for the tow upcoming weeks.

**Comment by florent-lamiraux** [comment]
I will have a look.

**Comment by florent-lamiraux** [comment]
I have successfully compiled and installed the project. However, I am not sure about what to do next.
Starting hpp-gui without hppcorbaserver running fails. Moreover I do not know what is the format requested when
loading a demo.

**Comment by ymontmarin** [comment]
Thank you @florent-lamiraux !
I did not choose to restore the hppcorbaserver plugin that launch the hppcorbaserver when launching the gui. I personally use an external app that launch both.

The demo file are using xml format giving joint bounds and set of named config. An example as follow:
```
<?xml version="1.0" ?>
<problem robotname="Voiture">
    <joint name="root_joint">-15 15 -15 15 -1.01 1.01 -1.01 1.01</joint>
    <joint name="wheel_frontleft_joint">-0.5 0.5</joint>
    <joint name="wheel_frontright_joint">-0.5 0.5</joint>
    <config name="parked">7.5 3.5 0.031790969350755503 -0.99949453938865385 0 0</config>
    <config name="couloir">-3 3.5 0.031790969350755503 -0.99949453938865385 0 0</config>
</problem>
```
But the use is more about playing with the gui (setting bounds, registering configuration) and saving it with `Save Demo` for future reuse with `Loading Demo` for fast demonstration.

Then `Reset` just allow to switch easilly between robot/env setup for demonstration.

Note that the demonstration does not include (maybe it should ?) the choices in the solver (algo, steering method...) nor the roadmap.

Overall the goal is to add tools to make hpp-gui easier for fast demonstration in the context of e.g. a lesson but still more interactive than a video!

**Comment by florent-lamiraux** [comment]
I think I miss something.
If I understand well, the demo file only stores
  * the name of the robot,
  * the joint bounds, and
  * some configurations.
Thus, starting a session still requires a python script to load the robot and obstacles from URDF files, right ?

**Comment by ymontmarin** [comment]
It is used jointly with the `Load Robot`, `Load Environment` dialog box (But it could be a python script.). I did not integrate this loading in the demo because not all of those information can be saved from a running session (eg some urdf adress...). Demo file purpose is to load what can be save from a previous session. (solver settings could be added..).

Again, the demo feature is only really useful for a specific need: improve user experience and rapidity to make demonstration without any coding, specially useful for lectures ! I don't use it on everyday experiments.

**Comment by florent-lamiraux** [comment]
Can you add a file explaining how to run a simple example ? I am unable to load a robot and to solve a problem.

File: plugins/hppwidgetsplugin/configurationlistwidget.cc [hunk 1]
```diff
@@ -61,23 +61,31 @@ namespace hpp {
...
 
     void ConfigurationListWidget::onSaveClicked ()
     {
       hpp::floatSeq const* c = plugin_->getCurrentConfig ();
       list()->addItem(makeItem(name()->text(), *c));
+      count_++;
       name()->setText(basename_ + QString::number(count_));
+    }
+
+    void ConfigurationListWidget::reciveConfig (QString name_, const hpp::floatSeq& config)
```
Comment (jmirabel) [inline plugins/hppwidgetsplugin/configurationlistwidget.cc]:
```suggestion
    void ConfigurationListWidget::receiveConfig (QString name_, const hpp::floatSeq& config)
```

File: plugins/hppwidgetsplugin/configurationlistwidget.hh [hunk 2]
```diff
@@ -46,13 +46,18 @@ namespace hpp {
 
         static hpp::floatSeq& getConfig (QListWidgetItem* item);
 
-        public slots:
-          /// Save the current configuration of the robot.
-          void onSaveClicked ();
+        // Added for external manipulation of the widget by demo widget
+        void reciveConfig (QString name, const hpp::floatSeq& config);
```
Comment (jmirabel) [inline plugins/hppwidgetsplugin/configurationlistwidget.hh]:
```suggestion
        void receiveConfig (QString name, const hpp::floatSeq& config);
```

**Review by jmirabel — APPROVED** [review]
As @florent-lamiraux suggested, a simple markdown file explaining how to use your widget and we are good to go.

**Comment by ymontmarin** [comment]
@jmirabel @florent-lamiraux 
Thank you for your responses and sorry for the delay, I agree that a markdown will be useful to explain the use of the feature ! I am actually doing one with some usage example for destination user and I will add it here when it is done.

**Pull request closed without merging**